File: src/lib/objmdl/mdl/mdl.ts
import { parsers } from '@pqhuy98/mdx-m3-viewer';

import { Vector3 } from '../../math/common';
import { Animation } from './components/animation';
import { Camera, camerasToString } from './components/camera';
import { Bound } from './components/extent';
import { f } from './components/formatter';
import {
  Geoset, GeosetAnim, geosetAnimsToString, geosetsToString,
} from './components/geoset';
import { GlobalSequence, globalSequencesToString } from './components/global-sequence';
import { Material, materialsToString } from './components/material';
import { Light, lightsToString } from './components/node/light';
import {
  AttachmentPoint, attachmentPointsToString, Bone, bonesToString, CollisionShape, collisionShapesToString, EventObject, eventObjectsToString, Helper, helpersToString, pivotPointsToString,
} from './components/node/node';
import { ParticleEmitter2, particleEmitter2sToString } from './components/node/particle-emitter-2';
import { Sequence, sequencesToString } from './components/sequence';
import { Texture, texturesToString } from './components/texture';
import { TextureAnim, textureAnimsToString } from './components/texture-anim';
import { MDLModify } from './modify';

export interface WowAttachment {
  wowAttachmentId: number;
  bone: Bone;
  pivotPoint: Vector3;
}

export class MDL {
  version: {
    formatVersion: number;
  };

  model: {
    name: string;
    blendTime: number;
  } & Bound;

  globalSequences: GlobalSequence[] = [];

  sequences: Sequence[] = [];

  textures: Texture[] = [];

  materials: Material[] = [];

  textureAnims: TextureAnim[] = [];

  geosets: Geoset[] = [];

  geosetAnims: GeosetAnim[] = [];

  bones: Bone[] = [];

  attachments: AttachmentPoint[] = [];

  lights: Light[] = [];

  // particleEmitters: ParticleEmitter[] = [];
  particleEmitter2s: ParticleEmitter2[] = [];
  // particleEmitterPopcorns: ParticleEmitterPopcorn[] = [];

  cameras: Camera[] = [];

  eventObjects: EventObject[] = [];

  helpers: Helper[] = [];

  collisionShapes: CollisionShape[] = [];

  modify: MDLModify;

  boundsOverriden?: (obj: Bound) => void;

  wowAttachments: WowAttachment[] = [];

  constructor(props: {formatVersion: number, name: string}) {
    this.version = { formatVersion: props.formatVersion };
    this.model = {
      name: props.name,
      blendTime: 150,
      minimumExtent: [0, 0, 0],
      maximumExtent: [0, 0, 0],
      boundsRadius: 0,
    };
    this.modify = new MDLModify(this);
  }

  private versionToString() {
    return `
      Version {
        FormatVersion ${this.version.formatVersion},
      }`;
  }

  private modelToString() {
    return `Model "${this.model.name}" {
      NumGeosets ${this.geosets.length},
      NumBones ${this.bones.length},
      NumAttachments ${this.attachments.length},
      BlendTime ${this.model.blendTime},
      MinimumExtent { ${this.model.minimumExtent.map(f).join(', ')} },
      MaximumExtent { ${this.model.maximumExtent.map(f).join(', ')} },
      BoundsRadius ${f(this.model.boundsRadius)},
    }`;
  }

  getNodes() {
    return [
      ...this.bones,
      ...this.attachments,
      ...this.lights,
      // ...this.particleEmitters,
      ...this.particleEmitter2s,
      // ...this.particleEmitterPopcorns,
      ...this.eventObjects,
      ...this.collisionShapes,
      ...this.helpers,
    ];
  }

  getAnimated(): Animation<number[] | number>[] {
    return [
      ...this.getNodes().flatMap((node) => [node.translation, node.rotation, node.scaling]),
      ...this.cameras.flatMap((cam) => [cam.translation, cam.rotation, cam.scaling]),
      ...this.textureAnims.flatMap((texAnim) => [texAnim.translation, texAnim.rotation, texAnim.scaling]),
      ...this.materials.flatMap((mat) => mat.layers.flatMap((layer) => [
        layer.alpha && 'keyFrames' in layer.alpha ? layer.alpha : null,
        layer.tvertexAnim?.translation,
        layer.tvertexAnim?.rotation,
        layer.tvertexAnim?.scaling,
      ])),
      ...this.geosetAnims.flatMap((geosetAnim) => [
        geosetAnim.alpha && 'keyFrames' in geosetAnim.alpha ? geosetAnim.alpha : null,
        geosetAnim.color && 'keyFrames' in geosetAnim.color ? geosetAnim.color : null,
      ]),
      ...this.lights.flatMap((l) => [
        l.visibility,
        'keyFrames' in l.attenuationStart ? l.attenuationStart : null,
        'keyFrames' in l.attenuationEnd ? l.attenuationEnd : null,
        'keyFrames' in l.intensity ? l.intensity : null,
        'keyFrames' in l.color ? l.color : null,
        'keyFrames' in l.ambientIntensity ? l.ambientIntensity : null,
        'keyFrames' in l.ambientColor ? l.ambientColor : null,
      ]),
      ...this.particleEmitter2s.flatMap((p) => [
        p.visibility,
        p.emissionRate && 'keyFrames' in p.emissionRate ? p.emissionRate : null,
        p.latitude && 'keyFrames' in p.latitude ? p.latitude : null,
        p.speed && 'keyFrames' in p.speed ? p.speed : null,
        p.variation && 'keyFrames' in p.variation ? p.variation : null,
        p.gravity && 'keyFrames' in p.gravity ? p.gravity : null,
        p.width && 'keyFrames' in p.width ? p.width : null,
        p.length && 'keyFrames' in p.length ? p.length : null,
      ]),
    ].filter((anim) => anim != null);
  }

  updateIds() {
    // Reindex everything with `id`
    this.globalSequences.forEach((v, i) => v.id = i);
    this.textures.forEach((v, i) => v.id = i);
    this.materials.forEach((v, i) => v.id = i);
    this.textureAnims.forEach((v, i) => v.id = i);
    this.geosetAnims.forEach((v, i) => v.id = i);
    this.geosets.forEach((geoset, i) => {
      geoset.id = i;
    });
    this.getNodes().forEach((node, i) => node.objectId = i);
    this.attachments.forEach((p, i) => p.attachmentId = i);
  }

  private toString() {
    this.updateIds();

    // Override extends before write to String
    if (this.boundsOverriden) {
      this.boundsOverriden(this.model);
      this.geosets.forEach((geoset) => {
        this.boundsOverriden!(geoset);
      });
      this.sequences.forEach((seq) => {
        this.boundsOverriden!(seq);
      });
    }

    let result = `// Exported by Huy's wow-converter
      ${this.versionToString()}
      ${this.modelToString()}
      ${sequencesToString(this.sequences)}
      ${globalSequencesToString(this.globalSequences)}
      ${texturesToString(this.textures)}
      ${materialsToString(this.version.formatVersion, this.materials)}
      ${textureAnimsToString(this.textureAnims)}
      ${geosetsToString(this.version.formatVersion, this.geosets, this.bones, this.sequences)}
      ${geosetAnimsToString(this.geosetAnims)}
      ${bonesToString(this.bones)}
      ${attachmentPointsToString(this.attachments)}
      ${lightsToString(this.lights)}
      ${particleEmitter2sToString(this.particleEmitter2s)}
      ${camerasToString(this.cameras)}
      ${eventObjectsToString(this.eventObjects)}
      ${collisionShapesToString(this.collisionShapes)}
      ${helpersToString(this.helpers)}
      ${pivotPointsToString(this.getNodes())}
    `;

    let depth = 0;
    result = result.split('\n')
      .map((l) => l.trim())
      .filter((l) => l.length > 0)
      .map((l) => {
        if (l === '}' || l === '},') {
          depth--;
        }
        try {
          const newLine = Array(depth).fill('\t').join('') + l;
          if (l.endsWith('{')) {
            depth++;
          }
          return newLine;
        } catch (e) {
          return l;
        }
      })
      .join('\n');

    // Restore
    if (this.boundsOverriden) {
      this.syncExtents();
    }

    return result;
  }

  toMdx() {
    const m = new parsers.mdlx.Model();
    const mdlStr = this.toString();
    m.loadMdl(mdlStr);
    return m.saveMdx();
  }

  toMdl() {
    return this.toString();
  }

  sync() {
    this.syncExtents();

    // Update global sequence durations to max of all key-frames timestamps
    const updateGlobalSeq = (globalSeq?: GlobalSequence, ...values: number[]) => {
      if (globalSeq) {
        globalSeq.duration = Math.max(globalSeq.duration, ...values);
      }
    };
    this.textureAnims.forEach((texAnim) => {
      updateGlobalSeq(texAnim.translation?.globalSeq, ...texAnim.translation?.keyFrames.keys() ?? []);
      updateGlobalSeq(texAnim.rotation?.globalSeq, ...texAnim.rotation?.keyFrames.keys() ?? []);
      updateGlobalSeq(texAnim.scaling?.globalSeq, ...texAnim.scaling?.keyFrames.keys() ?? []);
    });
    this.bones.forEach((bone) => {
      updateGlobalSeq(bone.translation?.globalSeq, ...bone.translation?.keyFrames.keys() ?? []);
      updateGlobalSeq(bone.rotation?.globalSeq, ...bone.rotation?.keyFrames.keys() ?? []);
      updateGlobalSeq(bone.scaling?.globalSeq, ...bone.scaling?.keyFrames.keys() ?? []);
    });

    // Compute bone's GeosetId. It can be ID of one single geoset, or "Multiple" if
    // the bone or any of its children is shared between multiple geosets.
    const geosetsPerBone = new Map<Bone, Set<Geoset>>();
    const boneChildren = new Map<Bone, Bone[]>();

    // Compute individual bone's list of geosets
    this.geosets.forEach((geoset) => {
      const bones: Set<Bone> = new Set();
      geoset.matrices.forEach((m) => m.bones.forEach((b) => bones.add(b)));
      geoset.vertices.forEach((v) => v.skinWeights?.forEach((w) => bones.add(w.bone)));
      bones.forEach((bone) => {
        if (!geosetsPerBone.has(bone)) {
          geosetsPerBone.set(bone, new Set());
        }
        geosetsPerBone.get(bone)!.add(geoset);
      });
    });

    this.bones.forEach((bone) => {
      // Compute bone's geoset
      if (geosetsPerBone.has(bone)) {
        const geosets = geosetsPerBone.get(bone)!;
        bone.geoset = geosets.size > 1 ? 'Multiple' : geosets.values().next().value;
      }
      // its children list for dfs
      const parent = bone.parent;
      if (parent) {
        if (!boneChildren.has(parent)) {
          boneChildren.set(parent, []);
        }
      boneChildren.get(parent)!.push(bone);
      }
    });

    const dfs = (node: Bone) => {
      for (const child of boneChildren.get(node) ?? []) {
        dfs(child);
        if (child.geoset === 'Multiple') {
          node.geoset = 'Multiple';
        }
        if (child.geoset && node.geoset && child.geoset !== node.geoset) {
          node.geoset = 'Multiple';
        }
        if (child.geoset && !node.geoset) {
          node.geoset = child.geoset;
        }
      }
    };
    this.bones.forEach((bone) => {
      if (!bone.parent) {
        dfs(bone);
      }
    });

    // If no material is defined, create a default one
    if (this.materials.length === 0) {
      this.textures = [
        {
          id: 0,
          image: '',
          wrapWidth: false,
          wrapHeight: false,
          wowData: {
            type: 0,
          },
        },
      ];
      this.materials = [
        {
          id: 0,
          twoSided: false,
          layers: [
            {
              filterMode: 'None',
              texture: this.textures[0],
              unshaded: false,
              sphereEnvMap: false,
              twoSided: false,
              unfogged: false,
              unlit: false,
              noDepthTest: false,
              noDepthSet: false,
              alpha: {
                static: true,
                value: 1,
              },
            },
          ],
          constantColor: false,
        },
      ];
    }

    this.geosets.forEach((geoset) => {
      if (!geoset.material) {
        geoset.material = this.materials[0];
      }
    });
  }

  syncExtents() {
    // Remove empty geosets
    this.geosets = this.geosets.filter((geoset) => geoset.vertices.length > 0 && geoset.faces.length > 0);

    // Compute geoset's bound
    this.geosets.forEach((geoset) => {
      const min: Vector3 = [Infinity, Infinity, Infinity];
      const max: Vector3 = [-Infinity, -Infinity, -Infinity];
      geoset.vertices.forEach(({ position: [x, y, z] }) => {
        min[0] = Math.min(min[0], x);
        min[1] = Math.min(min[1], y);
        min[2] = Math.min(min[2], z);
        max[0] = Math.max(max[0], x);
        max[1] = Math.max(max[1], y);
        max[2] = Math.max(max[2], z);
      });
      geoset.minimumExtent = min;
      geoset.maximumExtent = max;
      geoset.boundsRadius = calculateBoundRadius(geoset.vertices.map((v) => v.position));
    });

    // Compute model's bound
    if (this.geosets.length > 0) {
      this.model.minimumExtent = this.geosets.map((geoset) => geoset.minimumExtent).reduce((a, b) => [Math.min(a[0], b[0]), Math.min(a[1], b[1]), Math.min(a[2], b[2])]);
      this.model.maximumExtent = this.geosets.map((geoset) => geoset.maximumExtent).reduce((a, b) => [Math.max(a[0], b[0]), Math.max(a[1], b[1]), Math.max(a[2], b[2])]);
      this.model.boundsRadius = this.geosets.reduce((a, b) => Math.max(a, b.boundsRadius), 0);
    }

    // Update sequence's bound
    this.sequences.forEach((s) => {
      s.minimumExtent = [...this.model.minimumExtent];
      s.maximumExtent = [...this.model.maximumExtent];
      s.boundsRadius = this.model.boundsRadius;
    });
  }
}

function calculateBoundRadius(vertices: Vector3[]) {
  let maxDistance = 0;
  for (let i = 0; i < vertices.length; i++) {
    const [x, y, z] = vertices[i];
    const distance = Math.sqrt(x * x + y * y + z * z);
    if (distance > maxDistance) {
      maxDistance = distance;
    }
  }

  return maxDistance;
}

File: golang/internal/formats/mdl/mdl.go
package mdl

import (
	"fmt"
	stdmath "math"
	"strings"

	"github.com/pqhuy98/wow-converter/internal/formats/mdl/components"
	"github.com/pqhuy98/wow-converter/internal/formats/mdx"
	imath "github.com/pqhuy98/wow-converter/internal/math"
)

type Version struct {
	FormatVersion int
}

type ModelInfo struct {
	components.Bound
	Name      string
	BlendTime int
}

type MDL struct {
	Version           Version
	Model             ModelInfo
	GlobalSequences   []*components.GlobalSequence
	Sequences         []components.Sequence
	Textures          []*components.Texture
	Materials         []*components.Material
	TextureAnims      []components.TextureAnim
	Geosets           []*components.Geoset
	GeosetAnims       []components.GeosetAnim
	Bones             []*components.Bone
	Attachments       []*components.AttachmentPoint
	Lights            []*components.Light
	ParticleEmitter2s []*components.ParticleEmitter2
	RibbonEmitters    []*components.RibbonEmitter
	Cameras           []components.Camera
	EventObjects      []*components.EventObject
	Helpers           []*components.Helper
	CollisionShapes   []*components.CollisionShape
	Modify            *Modify
	BoundsOverriden   func(obj *components.Bound)
	WowAttachments    []components.WowAttachment
	AccumScale        float64
}

type NewMDLOptions struct {
	FormatVersion int
	Name          string
}

func New(opts NewMDLOptions) *MDL {
	m := &MDL{
		Version: Version{FormatVersion: opts.FormatVersion},
		Model: ModelInfo{
			Name:      opts.Name,
			BlendTime: 150,
		},
		AccumScale: 1,
	}
	m.Modify = NewModify(m)
	return m
}

func (m *MDL) GetNodes() []components.Node {
	nodes := make([]components.Node, 0)
	for _, b := range m.Bones {
		nodes = append(nodes, b)
	}
	for _, a := range m.Attachments {
		nodes = append(nodes, a)
	}
	for _, l := range m.Lights {
		nodes = append(nodes, l)
	}
	for _, r := range m.RibbonEmitters {
		nodes = append(nodes, r)
	}
	for _, p := range m.ParticleEmitter2s {
		nodes = append(nodes, p)
	}
	for _, e := range m.EventObjects {
		nodes = append(nodes, e)
	}
	for _, c := range m.CollisionShapes {
		nodes = append(nodes, c)
	}
	for _, h := range m.Helpers {
		nodes = append(nodes, h)
	}
	return nodes
}

func (m *MDL) GetAnimated() []*components.Animation {
	seen := map[*components.Animation]struct{}{}
	var result []*components.Animation
	add := func(anim *components.Animation) {
		if anim == nil || len(anim.KeyFrames) == 0 {
			return
		}
		if _, ok := seen[anim]; ok {
			return
		}
		seen[anim] = struct{}{}
		result = append(result, anim)
	}
	addAnimOrStatic := func(v any) {
		add(components.GetAnim(v))
	}

	for _, node := range m.GetNodes() {
		add(node.NodeTranslation())
		add(node.NodeRotation())
		add(node.NodeScaling())
	}
	for _, cam := range m.Cameras {
		add(cam.Translation)
		add(cam.Rotation)
		add(cam.Scaling)
	}
	for _, texAnim := range m.TextureAnims {
		add(texAnim.Translation)
		add(texAnim.Rotation)
		add(texAnim.Scaling)
	}
	for _, mat := range m.Materials {
		for _, layer := range mat.Layers {
			addAnimOrStatic(layer.Alpha)
			if layer.TVertexAnim != nil {
				add(layer.TVertexAnim.Translation)
				add(layer.TVertexAnim.Rotation)
				add(layer.TVertexAnim.Scaling)
			}
		}
	}
	for _, geosetAnim := range m.GeosetAnims {
		addAnimOrStatic(geosetAnim.Alpha)
		addAnimOrStatic(geosetAnim.Color)
	}
	for _, l := range m.Lights {
		add(l.Visibility)
		addAnimOrStatic(l.AttenuationStart)
		addAnimOrStatic(l.AttenuationEnd)
		addAnimOrStatic(l.Intensity)
		addAnimOrStatic(l.Color)
		addAnimOrStatic(l.AmbientIntensity)
		addAnimOrStatic(l.AmbientColor)
	}
	for _, r := range m.RibbonEmitters {
		add(r.Visibility)
		if r.HeightAbove != nil {
			addAnimOrStatic(r.HeightAbove)
		}
		if r.HeightBelow != nil {
			addAnimOrStatic(r.HeightBelow)
		}
		if r.Alpha != nil {
			addAnimOrStatic(r.Alpha)
		}
		if r.Color != nil {
			addAnimOrStatic(r.Color)
		}
		if r.TextureSlot != nil {
			addAnimOrStatic(r.TextureSlot)
		}
	}
	for _, p := range m.ParticleEmitter2s {
		add(p.Visibility)
		addAnimOrStatic(p.EmissionRate)
		addAnimOrStatic(p.Latitude)
		addAnimOrStatic(p.Speed)
		addAnimOrStatic(p.Variation)
		addAnimOrStatic(p.Gravity)
		addAnimOrStatic(p.Width)
		addAnimOrStatic(p.Length)
	}
	return result
}

func (m *MDL) UpdateIDs() {
	for i, gs := range m.GlobalSequences {
		gs.ID = i
	}
	for i := range m.Textures {
		m.Textures[i].ID = i
	}
	textureIDs := map[string]int{}
	for i := range m.Textures {
		key := textureUpdateKey(*m.Textures[i])
		if _, ok := textureIDs[key]; !ok {
			textureIDs[key] = i
		}
	}
	syncTextureID := func(texture *components.Texture) {
		if texture == nil {
			return
		}
		if id, ok := textureIDs[textureUpdateKey(*texture)]; ok {
			texture.ID = id
		}
	}
	for i := range m.Materials {
		m.Materials[i].ID = i
		for li := range m.Materials[i].Layers {
			syncTextureID(m.Materials[i].Layers[li].Texture)
		}
	}
	for _, emitter := range m.ParticleEmitter2s {
		syncTextureID(emitter.Texture)
	}
	for i := range m.TextureAnims {
		m.TextureAnims[i].ID = i
	}
	for i := range m.GeosetAnims {
		m.GeosetAnims[i].ID = i
	}
	for i, geoset := range m.Geosets {
		geoset.ID = i
	}
	nodes := m.GetNodes()
	for i, node := range nodes {
		node.SetNodeObjectID(i)
	}
	for i := range m.Attachments {
		m.Attachments[i].AttachmentID = i
	}
}

func textureUpdateKey(texture components.Texture) string {
	replaceableID := 0
	hasReplaceableID := false
	if texture.ReplaceableID != nil {
		replaceableID = *texture.ReplaceableID
		hasReplaceableID = true
	}
	return fmt.Sprintf(
		"%t|%d|%s|%t|%t|%d|%s",
		hasReplaceableID,
		replaceableID,
		texture.Image,
		texture.WrapWidth,
		texture.WrapHeight,
		texture.WowData.Type,
		texture.WowData.PngPath,
	)
}

func (m *MDL) ToMdl() string {
	return m.formatString()
}

func (m *MDL) ToMdx() ([]byte, error) {
	return mdx.ConvertMdlToMdx(m.ToMdl())
}

func (m *MDL) Sync() {
	m.SyncExtents()

	updateGlobalSeq := func(globalSeq *components.GlobalSequence, values ...int) {
		if globalSeq == nil {
			return
		}
		for _, v := range values {
			if v > globalSeq.Duration {
				globalSeq.Duration = v
			}
		}
	}

	for _, texAnim := range m.TextureAnims {
		if texAnim.Translation != nil {
			updateGlobalSeq(texAnim.Translation.GlobalSeq, components.SortedKeyInts(texAnim.Translation.KeyFrames)...)
		}
		if texAnim.Rotation != nil {
			updateGlobalSeq(texAnim.Rotation.GlobalSeq, components.SortedKeyInts(texAnim.Rotation.KeyFrames)...)
		}
		if texAnim.Scaling != nil {
			updateGlobalSeq(texAnim.Scaling.GlobalSeq, components.SortedKeyInts(texAnim.Scaling.KeyFrames)...)
		}
	}
	for _, bone := range m.Bones {
		if bone.Translation != nil {
			updateGlobalSeq(bone.Translation.GlobalSeq, components.SortedKeyInts(bone.Translation.KeyFrames)...)
		}
		if bone.Rotation != nil {
			updateGlobalSeq(bone.Rotation.GlobalSeq, components.SortedKeyInts(bone.Rotation.KeyFrames)...)
		}
		if bone.Scaling != nil {
			updateGlobalSeq(bone.Scaling.GlobalSeq, components.SortedKeyInts(bone.Scaling.KeyFrames)...)
		}
	}

	geosetsPerBone := map[*components.Bone]map[*components.Geoset]struct{}{}
	boneChildren := map[*components.Bone][]*components.Bone{}

	for _, geoset := range m.Geosets {
		bones := map[*components.Bone]struct{}{}
		for _, matrix := range geoset.Matrices {
			for _, b := range matrix.Bones {
				bones[b] = struct{}{}
			}
		}
		for _, v := range geoset.Vertices {
			for _, w := range v.SkinWeights {
				bones[w.Bone] = struct{}{}
			}
		}
		for bone := range bones {
			if geosetsPerBone[bone] == nil {
				geosetsPerBone[bone] = map[*components.Geoset]struct{}{}
			}
			geosetsPerBone[bone][geoset] = struct{}{}
		}
	}

	for _, bone := range m.Bones {
		if geosets, ok := geosetsPerBone[bone]; ok {
			if len(geosets) > 1 {
				bone.GeosetMulti = true
				bone.Geoset = nil
			} else {
				for g := range geosets {
					bone.Geoset = g
					bone.GeosetMulti = false
				}
			}
		}
		if bone.ParentBone != nil {
			boneChildren[bone.ParentBone] = append(boneChildren[bone.ParentBone], bone)
		} else if bone.Parent != nil {
			if pb, ok := bone.Parent.(*components.Bone); ok {
				boneChildren[pb] = append(boneChildren[pb], bone)
			}
		}
	}

	var dfs func(node *components.Bone)
	dfs = func(node *components.Bone) {
		for _, child := range boneChildren[node] {
			dfs(child)
			if child.GeosetMulti {
				node.GeosetMulti = true
				node.Geoset = nil
			}
			if child.Geoset != nil && node.Geoset != nil && child.Geoset != node.Geoset {
				node.GeosetMulti = true
				node.Geoset = nil
			}
			if child.Geoset != nil && node.Geoset == nil && !node.GeosetMulti {
				node.Geoset = child.Geoset
			}
		}
	}
	for _, bone := range m.Bones {
		if bone.ParentBone == nil && bone.Parent == nil {
			dfs(bone)
		}
	}

	needsMaterial := false
	for _, geoset := range m.Geosets {
		if geoset.Material == nil {
			needsMaterial = true
			break
		}
	}
	if needsMaterial {
		if len(m.Materials) == 0 {
			tex := &components.Texture{
				ID:      0,
				Image:   "",
				WowData: components.TextureWowData{Type: 0, PngPath: ""},
			}
			m.Textures = append(m.Textures, tex)
			m.Materials = []*components.Material{{
				ID:            0,
				TwoSided:      false,
				ConstantColor: false,
				Layers: []components.Layer{{
					FilterMode: components.BlendNone,
					Texture:    tex,
					Unshaded:   false, SphereEnvMap: false, TwoSided: false,
					Unfogged: false, Unlit: false, NoDepthTest: false, NoDepthSet: false,
					Alpha: components.AnimatedOrStatic[float64]{Static: true, Value: 1},
				}},
			}}
		}
		for _, geoset := range m.Geosets {
			if geoset.Material == nil {
				geoset.Material = m.Materials[0]
			}
		}
	}
	m.UpdateIDs()
}

func (m *MDL) SyncExtents() {
	var filtered []*components.Geoset
	for _, geoset := range m.Geosets {
		if len(geoset.Vertices) > 0 && len(geoset.Faces) > 0 {
			filtered = append(filtered, geoset)
		}
	}
	m.Geosets = filtered

	for _, geoset := range m.Geosets {
		min := imath.Vector3{stdmath.Inf(1), stdmath.Inf(1), stdmath.Inf(1)}
		max := imath.Vector3{stdmath.Inf(-1), stdmath.Inf(-1), stdmath.Inf(-1)}
		for _, v := range geoset.Vertices {
			x, y, z := v.Position[0], v.Position[1], v.Position[2]
			if x < min[0] {
				min[0] = x
			}
			if y < min[1] {
				min[1] = y
			}
			if z < min[2] {
				min[2] = z
			}
			if x > max[0] {
				max[0] = x
			}
			if y > max[1] {
				max[1] = y
			}
			if z > max[2] {
				max[2] = z
			}
		}
		geoset.MinimumExtent = min
		geoset.MaximumExtent = max
		positions := make([]imath.Vector3, len(geoset.Vertices))
		for i, v := range geoset.Vertices {
			positions[i] = v.Position
		}
		geoset.BoundsRadius = calculateBoundRadius(positions)
	}

	if len(m.Geosets) > 0 {
		modelMin := m.Geosets[0].MinimumExtent
		modelMax := m.Geosets[0].MaximumExtent
		maxRadius := m.Geosets[0].BoundsRadius
		for _, geoset := range m.Geosets[1:] {
			modelMin = imath.V3Min(modelMin, geoset.MinimumExtent)
			modelMax = imath.V3Max(modelMax, geoset.MaximumExtent)
			if geoset.BoundsRadius > maxRadius {
				maxRadius = geoset.BoundsRadius
			}
		}
		m.Model.MinimumExtent = modelMin
		m.Model.MaximumExtent = modelMax
		m.Model.BoundsRadius = maxRadius
	}

	for i := range m.Sequences {
		m.Sequences[i].MinimumExtent = components.CopyVector3(m.Model.MinimumExtent)
		m.Sequences[i].MaximumExtent = components.CopyVector3(m.Model.MaximumExtent)
		m.Sequences[i].BoundsRadius = m.Model.BoundsRadius
	}
}

func (m *MDL) formatString() string {
	m.UpdateIDs()

	if m.BoundsOverriden != nil {
		m.BoundsOverriden(&m.Model.Bound)
		for _, geoset := range m.Geosets {
			m.BoundsOverriden(&geoset.Bound)
		}
		for i := range m.Sequences {
			m.BoundsOverriden(&m.Sequences[i].Bound)
		}
	}

	result := "// Exported by Huy's wow-converter\n" +
		m.versionToString() + "\n" +
		m.modelToString() + "\n" +
		components.SequencesToString(m.Sequences) + "\n" +
		components.GlobalSequencesToString(m.GlobalSequences) + "\n" +
		components.TexturesToString(m.Textures) + "\n" +
		components.MaterialsToString(m.Version.FormatVersion, m.Materials) + "\n" +
		components.TextureAnimsToString(m.TextureAnims) + "\n" +
		components.GeosetsToString(m.Version.FormatVersion, m.Geosets, m.Bones, m.Sequences) + "\n" +
		components.GeosetAnimsToString(m.GeosetAnims) + "\n" +
		components.BonesToString(m.Bones) + "\n" +
		components.AttachmentPointsToString(m.Attachments) + "\n" +
		components.LightsToString(m.Lights) + "\n" +
		components.RibbonEmittersToString(m.RibbonEmitters) + "\n" +
		components.ParticleEmitter2sToString(m.ParticleEmitter2s) + "\n" +
		components.CamerasToString(m.Cameras) + "\n" +
		components.EventObjectsToString(m.EventObjects) + "\n" +
		components.CollisionShapesToString(m.CollisionShapes) + "\n" +
		components.HelpersToString(m.Helpers) + "\n" +
		components.PivotPointsToString(m.GetNodes())

	result = indentMDLLikeTS(result)

	if m.BoundsOverriden != nil {
		m.SyncExtents()
	}
	return result
}

func (m *MDL) versionToString() string {
	return "Version {\nFormatVersion " + components.FVal(float64(m.Version.FormatVersion)) + ",\n}"
}

func (m *MDL) modelToString() string {
	return "Model \"" + m.Model.Name + "\" {\n" +
		"NumGeosets " + components.FVal(float64(len(m.Geosets))) + ",\n" +
		"NumBones " + components.FVal(float64(len(m.Bones))) + ",\n" +
		"NumAttachments " + components.FVal(float64(len(m.Attachments))) + ",\n" +
		"BlendTime " + components.FVal(float64(m.Model.BlendTime)) + ",\n" +
		"MinimumExtent { " + components.FVector3(m.Model.MinimumExtent) + " },\n" +
		"MaximumExtent { " + components.FVector3(m.Model.MaximumExtent) + " },\n" +
		"BoundsRadius " + components.FVal(m.Model.BoundsRadius) + ",\n" +
		"}"
}

func indentMDLLikeTS(input string) string {
	lines := strings.Split(input, "\n")
	trimmed := make([]string, 0, len(lines))
	for _, line := range lines {
		l := strings.TrimSpace(line)
		if l == "" {
			continue
		}
		trimmed = append(trimmed, l)
	}
	depth := 0
	out := make([]string, 0, len(trimmed))
	for _, l := range trimmed {
		if l == "}" || l == "}," {
			depth--
		}
		out = append(out, strings.Repeat("\t", depth)+l)
		if strings.HasSuffix(l, "{") {
			depth++
		}
	}
	return strings.Join(out, "\n")
}

func calculateBoundRadius(vertices []imath.Vector3) float64 {
	maxDistance := 0.0
	for _, v := range vertices {
		distance := stdmath.Sqrt(v[0]*v[0] + v[1]*v[1] + v[2]*v[2])
		if distance > maxDistance {
			maxDistance = distance
		}
	}
	return maxDistance
}
